libsolidity/analysis/TypeChecker.cpp

```diff
@@ -596,19 +596,32 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	if (isStructMemberDeclaration)
 		return false;
 
-	if (auto referenceType = dynamic_cast<ReferenceType const*>(varType))
+	auto compositeType = dynamic_cast<CompositeType const*>(varType);
+	if (compositeType == nullptr)
+		return false;
+
+	bool callDataCheckRequired = true;
```

Comment:
Not sure how accurately `callDataCheckRequired` is calculated.

Reply:
I did a bit of code archeology to figure out the original intention of this check. The commit where it was added is f6d1cee06b28f6af772e66e2aa6e4b403aef2530 in PR #8665.

There's also this bit from @ekpyron about the motivation behind it: https://github.com/ethereum/solidity/pull/8665#discussion_r409604385
> Updated it - the base type here is not necessarily "wrong", if it actually lives in memory (since that'll use tail pointers), but as public function argument it also has to be valid for calldata, so I started to check both for public function arguments in the type checker.

So looks like the assumption was just too broad and your changes go in a good direction but we probably have to figure out from scratch which kinds of parameters should be exempt from this check. Preferably we should have test cases that cover all of them.

Reply:
Hm... I guess it would/could be:
``bool callDataCheckRequired = _variable.annotation().type->dataStoredIn(DataLocation::CallData) || (_variable.annotation().type->dataStoredIn(DataLocation::Memory) && (_variable.isConstructorParameter() || (_variable.isCallableOrCatchParameter() && !_variable.isIndexed()));``